File: sdk/nodejs/src/nano/impl/nano-service-impl.ts
import {Client} from "../../client";
import {NanoService} from "../nano-service";
import {Code, CodeEntityInfo, Language} from "../model/code";
import {Repository} from "../../repository";
import {decodeBase64} from "../../util/base64";

export class NanoServiceImpl implements NanoService {
    private repo: Repository<Code>;

    public constructor(private client: Client) {
        this.repo = client.repo<Code>(CodeEntityInfo)
    }

    async deployCode(code: Code, override?: boolean): Promise<Code> {
        if (override) {
            return this.repo.apply(code)
        } else {
            return this.repo.create(code)
        }
    }

    deploy(name: string, language: Language, source: string, override?: boolean): Promise<Code> {
        return this.deployCode({
            name: name,
            language: language,
            content: decodeBase64(source),
        } as Code, override)
    }

    unDeploy(code: Code): Promise<Code> {
        return this.repo.delete(code.id)
    }


}

File: sdk/java/src/main/java/io/apibrew/client/nano/impl/NanoServiceImpl.java
package io.apibrew.client.nano.impl;

import io.apibrew.client.Client;
import io.apibrew.client.EntityInfo;
import io.apibrew.client.nano.model.Code;
import io.apibrew.client.nano.NanoService;
import lombok.RequiredArgsConstructor;

import java.util.Base64;

@RequiredArgsConstructor
public class NanoServiceImpl implements NanoService {
    private final Client client;

    @Override
    public Code deploy(Code code) {
        return client.repo(Code.class).create(code);
    }

    @Override
    public Code deploy(String name, Code.Language language, String source) {
        return deploy(name, language, source, false);
    }

    @Override
    public Code deploy(String name, Code.Language language, String source, boolean override) {
        return deploy(new Code().withName(name).withLanguage(language).withContent(Base64.getEncoder().encodeToString(source.getBytes())), override);
    }

    @Override
    public Code deploy(Code code, boolean override) {
        if (override) {
            return client.repo(Code.class).apply(code);
        } else {
            return client.repo(Code.class).create(code);
        }
    }

    @Override
    public Code unDeploy(Code code) {
        if (code.getId() != null) {
            return client.repo(Code.class).delete(code.getId().toString());
        } else {
            Code loadedCode = client.loadRecord(EntityInfo.fromEntityClass(Code.class), new Code().withName(code.getName()));

            return client.repo(Code.class).delete(loadedCode.getId().toString());
        }
    }
}
